File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/GRPC/MutationGRPCService.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// GRPC/MutationGRPCService.swift
//
// gRPC service stub for generic mutation operations.
// =============================================================================

import Foundation
import GRPC

// MARK: - Response Models

/// Mutation response data.
public struct MutationResponse: Sendable {
    public let data: Any?
    public let metadata: [String: Any]?
    
    public init(data: Any?, metadata: [String: Any]?) {
        self.data = data
        self.metadata = metadata
    }
}

// MARK: - Service

/// gRPC service stub for generic mutation operations.
/// Wraps the Grey Mutation gRPC service.
public final class MutationGRPCService: Sendable {
    
    private let channel: GreyChannel
    
    // MARK: - Initialization
    
    public init(channel: GreyChannel) {
        self.channel = channel
    }
    
    // MARK: - Operations
    
    /// Execute a generic mutation.
    /// - Parameters:
    ///   - accessToken: Current access token
    ///   - endpoint: The mutation endpoint
    ///   - method: HTTP method (POST, PUT, PATCH, DELETE)
    ///   - body: Optional request body
    /// - Returns: MutationResponse on success
    /// - Throws: GreyError on failure
    public func mutate(
        accessToken: String,
        endpoint: String,
        method: String = "POST",
        body: Any?
    ) async throws -> MutationResponse {
        do {
            let grpcChannel = channel.getChannel()
            let options = channel.callOptions(accessToken: accessToken)
            
            // Note: In production, call actual gRPC stub
            // let client = Mutation_MutationServiceAsyncClient(channel: grpcChannel)
            // let request = Mutation_MutationRequest.with {
            //     $0.endpoint = endpoint
            //     $0.method = method
            //     if let body = body {
            //         $0.body = serializeBody(body)
            //     }
            // }
            // let response = try await client.mutate(request, callOptions: options)
            
            return MutationResponse(
                data: nil,
                metadata: nil
            )
        } catch let error as GRPCStatus {
            throw GreyError.fromGRPCStatus(error)
        } catch {
            throw GreyError.fromError(error)
        }
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/grpc/MutationGrpcService.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// grpc/MutationGrpcService.kt
//
// gRPC service stub for generic mutation operations.
// =============================================================================

package com.grey.sdk.grpc

import com.grey.sdk.config.GreyOptions
import com.grey.sdk.error.GreyError
import com.grey.sdk.error.GreyResult
import io.grpc.StatusException
import io.grpc.StatusRuntimeException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Mutation request data.
 */
data class MutationRequest(
    val endpoint: String,
    val method: String = "POST",
    val body: Any? = null
)

/**
 * Mutation response data.
 * Uses Any to support dynamic response shapes.
 */
data class MutationResponse(
    val data: Any?,
    val metadata: Map<String, Any>? = null
)

/**
 * gRPC service stub for generic mutation operations.
 * Wraps the Grey Mutation gRPC service.
 */
class MutationGrpcService(
    private val channel: GreyChannel,
    private val options: GreyOptions
) {
    /**
     * Execute a generic mutation.
     *
     * @param accessToken Current access token
     * @param endpoint The mutation endpoint
     * @param method HTTP method (POST, PUT, PATCH, DELETE)
     * @param body Optional request body
     * @return GreyResult containing MutationResponse or error
     */
    suspend fun mutate(
        accessToken: String,
        endpoint: String,
        method: String = "POST",
        body: Any? = null
    ): GreyResult<MutationResponse> {
        return withContext(Dispatchers.IO) {
            try {
                val grpcChannel = channel.getChannel()
                
                // Note: In production, call actual gRPC stub with auth metadata
                // val metadata = channel.createAuthMetadata(accessToken)
                // val stub = MutationServiceGrpcKt.MutationServiceCoroutineStub(grpcChannel)
                //     .withInterceptors(MetadataUtils.newAttachHeadersInterceptor(metadata))
                // val response = stub.mutate(mutationRequest {
                //     this.endpoint = endpoint
                //     this.method = method
                //     this.body = serializeBody(body)
                // })
                
                // Simulated response - replace with actual gRPC call
                GreyResult.success(
                    MutationResponse(
                        data = null, // From gRPC response
                        metadata = null
                    )
                )
            } catch (e: StatusException) {
                GreyResult.failure(GreyError.fromGrpcStatus(e))
            } catch (e: StatusRuntimeException) {
                GreyResult.failure(GreyError.fromGrpcStatusRuntime(e))
            } catch (e: Throwable) {
                GreyResult.failure(GreyError.fromException(e))
            }
        }
    }
}
